### PR TITLE
New module to support bugzilla

### PR DESCRIPTION
**Pull request by xiangge**
##### SUMMARY
Adding the module to support bugzilla by REST API.
* Fetch bug by id
* Comment on bugs
* Change bugs' status
See the module's document for more details

##### ISSUE TYPE

 New Module Pull Request

##### COMPONENT NAME
 bugzilla.py
##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: lib/ansible/modules/web_infrastructure/bugzilla.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = """
+module: bugzilla
+version_added: "2.9"
+short_description: modify bugs in a Bugzilla instance by xmlrpc
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/bugzilla.py]:
```suggestion
short_description: Manage bugs in a Bugzilla instance by xmlrpc
```

File: lib/ansible/modules/web_infrastructure/bugzilla.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+description:
+  - Modifybugs in a Bugzilla instance.
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/bugzilla.py]:
```suggestion
  - This module can be used to fetch, comment, change status of bugzilla ids in a Bugzilla instance.
```

File: lib/ansible/modules/web_infrastructure/bugzilla.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+options:
+  uri:
+    required: true
+    description:
+      - Base URI for the Bugzilla instance.
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/bugzilla.py]:
```suggestion
      - Base URI for the Bugzilla instance.
    type: str
```

File: lib/ansible/modules/web_infrastructure/bugzilla.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+  operation:
+    required: true
+    aliases: [ command ]
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/bugzilla.py]:
```suggestion
    aliases: [ command ]
    type: str
```

File: lib/ansible/modules/web_infrastructure/bugzilla.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    choices: [ comment, transition, fetch ]
+    description:
+      - The operation to perform.
+
+  username:
+    required: true
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/bugzilla.py]:
```suggestion
    required: true
    type: str
```

File: lib/ansible/modules/web_infrastructure/bugzilla.py [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+    description:
+      - The username to log-in with.
+
+  password:
+    required: true
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/bugzilla.py]:
```suggestion
    required: true
    type: str
```

File: lib/ansible/modules/web_infrastructure/bugzilla.py [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+    description:
+      - The password to log-in with.
+
+  bug:
+    required: false
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/bugzilla.py]: type: str

File: lib/ansible/modules/web_infrastructure/bugzilla.py [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+    description:
+     - An existing bug number to operate on.
+
+  comment:
+    required: false
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/bugzilla.py]:
```suggestion
    required: false
    type: str
```

File: lib/ansible/modules/web_infrastructure/bugzilla.py [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+    uri: '{{ server }}'
+    username: '{{ user }}'
+    password: '{{ pass }}'
+    bug: '{{ bug.id }}'
+    operation: transition
+    status: ON_QA
+"""
+
+RETURN = '''
+bug:
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/bugzilla.py]:
```suggestion
bugzilla_info:
```

Comment (Akasurde) [reply lib/ansible/modules/web_infrastructure/bugzilla.py]: or ```bug_info```

File: lib/ansible/modules/web_infrastructure/bugzilla.py [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+
+# Handle request by ansible lib
+def handle_request(url, timeout, data=None, method='GET'):
+    data = json.dumps(data)
+    response, info = fetch_url(module, url, data=data, method=method, timeout=timeout,
+                               headers={'Content-Type': 'application/json'})
+    if info['status'] not in (200, 201, 204):
+        module.fail_json(msg=info['msg'])
+    body = response.read()
+    if body:
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/bugzilla.py]:
```suggestion
    ret = {}
    if body:
```

File: lib/ansible/modules/web_infrastructure/bugzilla.py [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+        return json.loads(body)
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/bugzilla.py]:
```suggestion
        ret = json.loads(body)
```

File: lib/ansible/modules/web_infrastructure/bugzilla.py [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+    else:
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/bugzilla.py]:
```suggestion
    return ret
```

File: lib/ansible/modules/web_infrastructure/bugzilla.py [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+            timeout=dict(type='float', default=10),
+        ),
+        supports_check_mode=False
+    )
+
+    op = module.params['operation']
+
+    # Check we have the necessary per-operation parameters
+    missing = []
+    for parm in OP_REQUIRED[op]:
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/bugzilla.py]: You can take a look at [require_if](https://github.com/ansible/ansible/blob/0ada6b8d8d070f8f1eb36cd82a1e4198a53e0b58/lib/ansible/modules/cloud/amazon/aws_config_recorder.py#L156) example.

**Review by Akasurde — CHANGES_REQUESTED** [review]
Would love to see unit test for this new module. Thanks for the PR.

**Comment by Akasurde** [comment]
@xiangge Are you still working on this, let us know ? Thanks, needs_info

**Comment by xiangge** [comment]
> yeah, still.
 Previously I was busy on other stuff.
 Now I want to continue this coding, will try to commit the new patch recently, hope it can pass the ci stuff, thanks for asking this.

File: lib/ansible/modules/web_infrastructure/bugzilla.py (hunk 7 above)
Comment (xiangge) [reply lib/ansible/modules/web_infrastructure/bugzilla.py]: Ack

File: lib/ansible/modules/web_infrastructure/bugzilla.py (hunk 13 above)
Comment (xiangge) [reply lib/ansible/modules/web_infrastructure/bugzilla.py]: Is that ok to follow the style of jira.py as I am using the AnsibleModule here?

**Comment by xiangge** [comment]
> Would love to see unit test for this new module. Thanks for the PR.
Yeah, will add test, thanks.

**Comment by xiangge** [comment]
> Would love to see unit test for this new module. Thanks for the PR.

May I ask a question here?
When I execute below commands, it raised "TypeError: Can't compile non template nodes", do you know why?

ansible-test sanity -v --venv --python 3.5 bugzilla
ansible-test bugzilla

```
Traceback (most recent call last):
  File "/usr/bin/ansible-test", line 69, in <module>
    main(context)
  File "/usr/bin/ansible-test", line 38, in main
    generate_dockerfile(context)
  File "/usr/bin/ansible-test", line 17, in generate_dockerfile
    resource_string("ansible_test","resources/Dockerfile.j2")
  File "/home/xx/.local/lib/python3.5/site-packages/jinja2/environment.py", line 945, in __new__
    return env.from_string(source, template_class=cls)
  File "/home/xx/.local/lib/python3.5/site-packages/jinja2/environment.py", line 880, in from_string
    return cls.from_code(self, self.compile(source), globals, None)
  File "/home/x/.local/lib/python3.5/site-packages/jinja2/environment.py", line 581, in compile
    defer_init=defer_init)
  File "/home/xx/.local/lib/python3.5/site-packages/jinja2/environment.py", line 543, in _generate
    optimized=self.optimized)
  File "/home/xx/.local/lib/python3.5/site-packages/jinja2/compiler.py", line 78, in generate
    raise TypeError('Can\'t compile non template nodes')
TypeError: Can't compile non template nodes
```

**Comment by xiangge** [comment]
> Would love to see unit test for this new module. Thanks for the PR.
Hi Akasurde, 
Do you know why there isn't a web_infrastructure in devel branch?
How can I contribute the module to upstream then?

**Comment by Akasurde** [comment]
@xiangge Could you please move this PR to https://github.com/ansible-collections/community.general since there was a separate collection for web_infrastructure?

**Comment by bcoca** [comment]
Hi!

Thank you very much for your submission to Ansible. It sincerely means a lot to us that you've taken the time to contribute.

Per recent announcements at AnsibleFest Atlanta 2019, going forward we are no longer accepting new one off modules/plugins, or modules/plugins that create a new category.

However, we recommend looking into providing this functionality through Ansible Galaxy via Ansible Collections. You can find more information about collections at https://docs.ansible.co
m/ansible/devel/dev_guide/developing_collections.html.

Thank you once again for this and your interest in Ansible!

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project

**Comment by xiangge** [comment]
> @xiangge Could you please move this PR to https://github.com/ansible-collections/community.general since there was a separate collection for web_infrastructure?

Thanks to the response, will do

**Pull request closed without merging**